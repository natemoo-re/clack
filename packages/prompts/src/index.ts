export { isCancel, mockPrompt, setGlobalAliases } from '@clack/core';
export { default as confirm, type ConfirmOptions } from './prompts/confirm';
export {
	default as group,
	type PromptGroup,
	type PromptGroupAwaitedReturn,
	type PromptGroupOptions
} from './prompts/group';
export {
	default as groupMultiselect,
	type GroupMultiSelectOptions
} from './prompts/group-multiselect';
export { cancel, default as log, intro, outro, type LogMessageOptions } from './prompts/log';
export { default as multiselect, type MultiSelectOptions } from './prompts/multiselect';
export { default as note } from './prompts/note';
export { default as password, type PasswordOptions } from './prompts/password';
export { default as select } from './prompts/select';
export { default as selectKey } from './prompts/select-key';
export { default as spinner } from './prompts/spinner';
export { default as tasks, type Task } from './prompts/tasks';
export { default as text, type TextOptions } from './prompts/text';
export { type Option, type SelectOptions } from './utils';
