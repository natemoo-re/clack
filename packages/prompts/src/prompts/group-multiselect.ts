import { GroupMultiSelectPrompt } from '@clack/core';
import color from 'picocolors';
import {
	Option,
	symbol,
	S_BAR,
	S_BAR_END,
	S_CHECKBOX_ACTIVE,
	S_CHECKBOX_INACTIVE,
	S_CHECKBOX_SELECTED
} from '../utils';

export const opt = <TValue>(
	option: Option<TValue>,
	state:
		| 'inactive'
		| 'active'
		| 'selected'
		| 'active-selected'
		| 'group-active'
		| 'group-active-selected'
		| 'submitted'
		| 'cancelled',
	options: Option<TValue>[] = []
) => {
	const label = option.label ?? String(option.value);
	const isItem = typeof (option as any).group === 'string';
	const next = isItem && (options[options.indexOf(option) + 1] ?? { group: true });
	const isLast = isItem && (next as any).group === true;
	const prefix = isItem ? `${isLast ? S_BAR_END : S_BAR} ` : '';

	if (state === 'group-active') {
		return `${prefix}${color.cyan(S_CHECKBOX_ACTIVE)} ${color.dim(label)}`;
	} else if (state === 'group-active-selected') {
		return `${prefix}${color.green(S_CHECKBOX_SELECTED)} ${color.dim(label)}`;
	} else if (state === 'active') {
		return `${color.dim(prefix)}${color.cyan(S_CHECKBOX_ACTIVE)} ${label}${
			option.hint ? ` ${color.dim(`(${option.hint})`)}` : ''
		}`;
	} else if (state === 'active-selected') {
		return `${color.dim(prefix)}${color.green(S_CHECKBOX_SELECTED)} ${label}${
			option.hint ? ` ${color.dim(`(${option.hint})`)}` : ''
		}`;
	} else if (state === 'selected') {
		return `${color.dim(prefix)}${color.green(S_CHECKBOX_SELECTED)} ${color.dim(label)}`;
	} else if (state === 'cancelled') {
		return `${color.strikethrough(color.dim(label))}`;
	} else if (state === 'submitted') {
		return `${color.dim(label)}`;
	}
	return `${color.dim(prefix)}${color.dim(S_CHECKBOX_INACTIVE)} ${color.dim(label)}`;
};

export interface GroupMultiSelectOptions<TValue> {
	message: string;
	options: Record<string, Option<TValue>[]>;
	initialValues?: TValue[];
	required?: boolean;
	cursorAt?: TValue;
}

const groupMultiselect = <TValue>(opts: GroupMultiSelectOptions<TValue>) => {
	return new GroupMultiSelectPrompt({
		options: opts.options,
		initialValues: opts.initialValues,
		required: opts.required ?? true,
		cursorAt: opts.cursorAt,
		validate(selected: TValue[]) {
			if (this.required && selected.length === 0)
				return `Please select at least one option.\n${color.reset(
					color.dim(
						`Press ${color.gray(color.bgWhite(color.inverse(' space ')))} to select, ${color.gray(
							color.bgWhite(color.inverse(' enter '))
						)} to submit`
					)
				)}`;
		},
		render() {
			let title = `${color.gray(S_BAR)}\n${symbol(this.state)}  ${opts.message}\n`;

			switch (this.state) {
				case 'submit': {
					return `${title}${color.gray(S_BAR)}  ${this.options
						.filter(({ value }) => this.value.includes(value))
						.map((option) => opt(option, 'submitted'))
						.join(color.dim(', '))}`;
				}
				case 'cancel': {
					const label = this.options
						.filter(({ value }) => this.value.includes(value))
						.map((option) => opt(option, 'cancelled'))
						.join(color.dim(', '));
					return `${title}${color.gray(S_BAR)}  ${
						label.trim() ? `${label}\n${color.gray(S_BAR)}` : ''
					}`;
				}
				case 'error': {
					const footer = this.error
						.split('\n')
						.map((ln, i) =>
							i === 0
								? `${color.yellow(S_BAR_END)}  ${color.yellow(ln)}`
								: `${color.hidden('-')}  ${ln}`
						)
						.join('\n');
					return `${title}${color.yellow(S_BAR)}  ${this.options
						.map((option, i, options) => {
							const selected =
								this.value.includes(option.value) ||
								(option.group === true && this.isGroupSelected(`${option.value}`));
							const active = i === this.cursor;
							const groupActive =
								!active &&
								typeof option.group === 'string' &&
								this.options[this.cursor].value === option.group;
							if (groupActive) {
								return opt(option, selected ? 'group-active-selected' : 'group-active', options);
							}
							if (active && selected) {
								return opt(option, 'active-selected', options);
							}
							if (selected) {
								return opt(option, 'selected', options);
							}
							return opt(option, active ? 'active' : 'inactive', options);
						})
						.join(`\n${color.yellow(S_BAR)}  `)}\n${footer}\n`;
				}
				default: {
					return `${title}${color.cyan(S_BAR)}  ${this.options
						.map((option, i, options) => {
							const selected =
								this.value.includes(option.value) ||
								(option.group === true && this.isGroupSelected(`${option.value}`));
							const active = i === this.cursor;
							const groupActive =
								!active &&
								typeof option.group === 'string' &&
								this.options[this.cursor].value === option.group;
							if (groupActive) {
								return opt(option, selected ? 'group-active-selected' : 'group-active', options);
							}
							if (active && selected) {
								return opt(option, 'active-selected', options);
							}
							if (selected) {
								return opt(option, 'selected', options);
							}
							return opt(option, active ? 'active' : 'inactive', options);
						})
						.join(`\n${color.cyan(S_BAR)}  `)}\n${color.cyan(S_BAR_END)}\n`;
				}
			}
		},
	}).prompt() as Promise<TValue[] | symbol>;
};

export default groupMultiselect;
