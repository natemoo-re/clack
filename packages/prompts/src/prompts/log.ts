import color from 'picocolors';
import {
	S_BAR,
	S_BAR_END,
	S_BAR_START,
	S_ERROR,
	S_INFO,
	S_STEP_SUBMIT,
	S_SUCCESS,
	S_WARN
} from '../utils';

export const cancel = (message = '') => {
	process.stdout.write(`${color.gray(S_BAR_END)}  ${color.red(message)}\n\n`);
};

export const intro = (title = '') => {
	process.stdout.write(`${color.gray(S_BAR_START)}  ${title}\n`);
};

export const outro = (message = '') => {
	process.stdout.write(`${color.gray(S_BAR)}\n${color.gray(S_BAR_END)}  ${message}\n\n`);
};

export type LogMessageOptions = {
	symbol?: string;
};

const log = {
	message: (message = '', { symbol = color.gray(S_BAR) }: LogMessageOptions = {}) => {
		const parts = [`${color.gray(S_BAR)}`];
		if (message) {
			const [firstLine, ...lines] = message.split('\n');
			parts.push(`${symbol}  ${firstLine}`, ...lines.map((ln) => `${color.gray(S_BAR)}  ${ln}`));
		}
		process.stdout.write(`${parts.join('\n')}\n`);
	},
	info: (message: string) => {
		log.message(message, { symbol: color.blue(S_INFO) });
	},
	success: (message: string) => {
		log.message(message, { symbol: color.green(S_SUCCESS) });
	},
	step: (message: string) => {
		log.message(message, { symbol: color.green(S_STEP_SUBMIT) });
	},
	warn: (message: string) => {
		log.message(message, { symbol: color.yellow(S_WARN) });
	},
	/** alias for `log.warn()`. */
	warning: (message: string) => {
		log.warn(message);
	},
	error: (message: string) => {
		log.message(message, { symbol: color.red(S_ERROR) });
	},
};

//@ts-expect-error - Name prompt to validate package exports
log.name = 'log';

export default log;
