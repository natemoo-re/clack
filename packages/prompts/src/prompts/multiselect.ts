import { MultiSelectPrompt } from '@clack/core';
import color from 'picocolors';
import {
	limitOptions,
	Option,
	symbol,
	S_BAR,
	S_BAR_END,
	S_CHECKBOX_ACTIVE,
	S_CHECKBOX_INACTIVE,
	S_CHECKBOX_SELECTED
} from '../utils';

export const opt = <TValue>(
	option: Option<TValue>,
	state: 'inactive' | 'active' | 'selected' | 'active-selected' | 'submitted' | 'cancelled'
) => {
	const label = option.label ?? String(option.value);
	if (state === 'active') {
		return `${color.cyan(S_CHECKBOX_ACTIVE)} ${label}${
			option.hint ? ` ${color.dim(`(${option.hint})`)}` : ''
		}`;
	} else if (state === 'selected') {
		return `${color.green(S_CHECKBOX_SELECTED)} ${color.dim(label)}`;
	} else if (state === 'cancelled') {
		return `${color.strikethrough(color.dim(label))}`;
	} else if (state === 'active-selected') {
		return `${color.green(S_CHECKBOX_SELECTED)} ${label}${
			option.hint ? ` ${color.dim(`(${option.hint})`)}` : ''
		}`;
	} else if (state === 'submitted') {
		return `${color.dim(label)}`;
	}
	return `${color.dim(S_CHECKBOX_INACTIVE)} ${color.dim(label)}`;
};

export interface MultiSelectOptions<TValue> {
	message: string;
	options: Option<TValue>[];
	initialValues?: TValue[];
	maxItems?: number;
	required?: boolean;
	cursorAt?: TValue;
}

const multiselect = <TValue>(opts: MultiSelectOptions<TValue>) => {
	return new MultiSelectPrompt({
		options: opts.options,
		initialValues: opts.initialValues,
		required: opts.required ?? true,
		cursorAt: opts.cursorAt,
		validate(selected: TValue[]) {
			if (this.required && selected.length === 0)
				return `Please select at least one option.\n${color.reset(
					color.dim(
						`Press ${color.gray(color.bgWhite(color.inverse(' space ')))} to select, ${color.gray(
							color.bgWhite(color.inverse(' enter '))
						)} to submit`
					)
				)}`;
		},
		render() {
			let title = `${color.gray(S_BAR)}\n${symbol(this.state)}  ${opts.message}\n`;

			const styleOption = (option: Option<TValue>, active: boolean) => {
				const selected = this.value.includes(option.value);
				if (active && selected) {
					return opt(option, 'active-selected');
				}
				if (selected) {
					return opt(option, 'selected');
				}
				return opt(option, active ? 'active' : 'inactive');
			};

			switch (this.state) {
				case 'submit': {
					return `${title}${color.gray(S_BAR)}  ${this.options
						.filter(({ value }) => this.value.includes(value))
						.map((option) => opt(option, 'submitted'))
						.join(color.dim(', '))}`;
				}
				case 'cancel': {
					const label = this.options
						.filter(({ value }) => this.value.includes(value))
						.map((option) => opt(option, 'cancelled'))
						.join(color.dim(', '));
					return `${title}${color.gray(S_BAR)}  ${
						label.trim() ? `${label}\n${color.gray(S_BAR)}` : ''
					}`;
				}
				case 'error': {
					const footer = this.error
						.split('\n')
						.map((ln, i) =>
							i === 0
								? `${color.yellow(S_BAR_END)}  ${color.yellow(ln)}`
								: `${color.hidden('-')}  ${ln}`
						)
						.join('\n');
					return (
						title +
						color.yellow(S_BAR) +
						'  ' +
						limitOptions({
							options: this.options,
							cursor: this.cursor,
							maxItems: opts.maxItems,
							style: styleOption,
						}).join(`\n${color.yellow(S_BAR)}  `) +
						'\n' +
						footer +
						'\n'
					);
				}
				default: {
					return `${title}${color.cyan(S_BAR)}  ${limitOptions({
						options: this.options,
						cursor: this.cursor,
						maxItems: opts.maxItems,
						style: styleOption,
					}).join(`\n${color.cyan(S_BAR)}  `)}\n${color.cyan(S_BAR_END)}\n`;
				}
			}
		},
	}).prompt() as Promise<TValue[] | symbol>;
};

export default multiselect;
