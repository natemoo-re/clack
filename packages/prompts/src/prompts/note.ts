import color from 'picocolors';
import {
	strip,
	S_BAR,
	S_BAR_H,
	S_CONNECT_LEFT,
	S_CORNER_BOTTOM_RIGHT,
	S_CORNER_TOP_RIGHT,
	S_STEP_SUBMIT
} from '../utils';

const note = (message = '', title = '') => {
	const lines = `\n${message}\n`.split('\n');
	const titleLen = strip(title).length;
	const len =
		Math.max(
			lines.reduce((sum, ln) => {
				ln = strip(ln);
				return ln.length > sum ? ln.length : sum;
			}, 0),
			titleLen
		) + 2;
	const msg = lines
		.map(
			(ln) =>
				`${color.gray(S_BAR)}  ${color.dim(ln)}${' '.repeat(len - strip(ln).length)}${color.gray(
					S_BAR
				)}`
		)
		.join('\n');
	process.stdout.write(
		`${color.gray(S_BAR)}\n${color.green(S_STEP_SUBMIT)}  ${color.reset(title)} ${color.gray(
			S_BAR_H.repeat(Math.max(len - titleLen - 1, 1)) + S_CORNER_TOP_RIGHT
		)}\n${msg}\n${color.gray(S_CONNECT_LEFT + S_BAR_H.repeat(len + 2) + S_CORNER_BOTTOM_RIGHT)}\n`
	);
};

export default note;
