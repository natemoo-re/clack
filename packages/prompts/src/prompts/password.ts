import { PasswordPrompt } from '@clack/core';
import color from 'picocolors';
import { symbol, S_BAR, S_BAR_END, S_PASSWORD_MASK } from '../utils';

export interface PasswordOptions {
	message: string;
	mask?: string;
	validate?: (value: string) => string | void;
}

const password = (opts: PasswordOptions) => {
	return new PasswordPrompt({
		validate: opts.validate,
		mask: opts.mask ?? S_PASSWORD_MASK,
		render() {
			const title = `${color.gray(S_BAR)}\n${symbol(this.state)}  ${opts.message}\n`;
			const masked = this.valueWithCursor;

			switch (this.state) {
				case 'error':
					return `${title.trim()}\n${color.yellow(S_BAR)}  ${masked}\n${color.yellow(
						S_BAR_END
					)}  ${color.yellow(this.error)}\n`;
				case 'submit':
					return `${title}${color.gray(S_BAR)}  ${color.dim(masked)}`;
				case 'cancel':
					return `${title}${color.gray(S_BAR)}  ${
						masked ? color.strikethrough(color.dim(masked)) : ''
					}${masked ? '\n' + color.gray(S_BAR) : ''}`;
				default:
					return `${title}${color.cyan(S_BAR)}  ${masked}\n${color.cyan(S_BAR_END)}\n`;
			}
		},
	}).prompt() as Promise<string | symbol>;
};

export default password;
