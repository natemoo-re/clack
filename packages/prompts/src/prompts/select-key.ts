import { SelectKeyPrompt } from '@clack/core';
import color from 'picocolors';
import { Option, SelectOptions, symbol, S_BAR, S_BAR_END } from '../utils';

export const opt = <TValue>(
	option: Option<TValue>,
	state: 'inactive' | 'active' | 'selected' | 'cancelled' = 'inactive'
) => {
	const label = option.label ?? String(option.value);
	if (state === 'selected') {
		return `${color.dim(label)}`;
	} else if (state === 'cancelled') {
		return `${color.strikethrough(color.dim(label))}`;
	} else if (state === 'active') {
		return `${color.bgCyan(color.gray(` ${option.value} `))} ${label}${
			option.hint ? ` ${color.dim(`(${option.hint})`)}` : ''
		}`;
	}
	return `${color.gray(color.bgWhite(color.inverse(` ${option.value} `)))} ${label}${
		option.hint ? ` ${color.dim(`(${option.hint})`)}` : ''
	}`;
};

const selectKey = <TValue extends string>(opts: SelectOptions<TValue>) => {
	return new SelectKeyPrompt({
		options: opts.options,
		initialValue: opts.initialValue,
		render() {
			const title = `${color.gray(S_BAR)}\n${symbol(this.state)}  ${opts.message}\n`;

			switch (this.state) {
				case 'submit':
					return `${title}${color.gray(S_BAR)}  ${opt(
						this.options.find((opt) => opt.value === this.value)!,
						'selected'
					)}`;
				case 'cancel':
					return `${title}${color.gray(S_BAR)}  ${opt(this.options[0], 'cancelled')}\n${color.gray(
						S_BAR
					)}`;
				default: {
					return `${title}${color.cyan(S_BAR)}  ${this.options
						.map((option, i) => opt(option, i === this.cursor ? 'active' : 'inactive'))
						.join(`\n${color.cyan(S_BAR)}  `)}\n${color.cyan(S_BAR_END)}\n`;
				}
			}
		},
	}).prompt() as Promise<TValue | symbol>;
};

export default selectKey;
