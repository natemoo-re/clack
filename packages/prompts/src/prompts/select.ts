import { SelectPrompt } from '@clack/core';
import color from 'picocolors';
import {
	limitOptions,
	Option,
	SelectOptions,
	symbol,
	S_BAR,
	S_BAR_END,
	S_RADIO_ACTIVE,
	S_RADIO_INACTIVE
} from '../utils';

export const opt = <TValue>(
	option: Option<TValue>,
	state: 'inactive' | 'active' | 'selected' | 'cancelled'
) => {
	const label = option.label ?? String(option.value);
	switch (state) {
		case 'selected':
			return `${color.dim(label)}`;
		case 'active':
			return `${color.green(S_RADIO_ACTIVE)} ${label} ${
				option.hint ? color.dim(`(${option.hint})`) : ''
			}`;
		case 'cancelled':
			return `${color.strikethrough(color.dim(label))}`;
		default:
			return `${color.dim(S_RADIO_INACTIVE)} ${color.dim(label)}`;
	}
};

const select = <TValue>(opts: SelectOptions<TValue>) => {
	return new SelectPrompt({
		options: opts.options,
		initialValue: opts.initialValue,
		render() {
			const title = `${color.gray(S_BAR)}\n${symbol(this.state)}  ${opts.message}\n`;

			switch (this.state) {
				case 'submit':
					return `${title}${color.gray(S_BAR)}  ${opt(this.options[this.cursor], 'selected')}`;
				case 'cancel':
					return `${title}${color.gray(S_BAR)}  ${opt(
						this.options[this.cursor],
						'cancelled'
					)}\n${color.gray(S_BAR)}`;
				default:
					return `${title}${color.cyan(S_BAR)}  ${limitOptions({
						cursor: this.cursor,
						options: this.options,
						maxItems: opts.maxItems,
						style: (item, active) => opt(item, active ? 'active' : 'inactive'),
					}).join(`\n${color.cyan(S_BAR)}  `)}\n${color.cyan(S_BAR_END)}\n`;
			}
		},
	}).prompt() as Promise<TValue | symbol>;
};

export default select;
