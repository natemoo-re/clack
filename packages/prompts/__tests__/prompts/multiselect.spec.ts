import { mockPrompt, SelectPrompt } from '@clack/core';
import color from 'picocolors';
import { multiselect } from '../../src';
import { opt } from '../../src/prompts/multiselect';
import {
	Option,
	symbol,
	S_BAR,
	S_BAR_END,
	S_CHECKBOX_ACTIVE,
	S_CHECKBOX_INACTIVE,
	S_CHECKBOX_SELECTED
} from '../../src/utils';

const options: Option<string>[] = [{ value: 'foo' }, { value: 'bar' }, { value: 'baz' }];

describe('multiselect', () => {
	const mock = mockPrompt<SelectPrompt<{ value: string }>>();
	const message = 'test message';

	afterEach(() => {
		mock.close();
	});

	it('should format option state with hint', () => {
		const option: Option<string> = {
			label: 'Foo',
			value: 'foo',
			hint: 'bar',
		};

		multiselect({ message, options });
		const hint = color.dim(`(${option.hint})`);

		expect(opt(option, 'active')).toBe(`${color.cyan(S_CHECKBOX_ACTIVE)} ${option.label} ${hint}`);
		expect(opt(option, 'active-selected')).toBe(
			`${color.green(S_CHECKBOX_SELECTED)} ${option.label} ${hint}`
		);
	});

	it('should format option state without hint', () => {
		const option: Option<string> = {
			label: 'Foo',
			value: 'foo',
		};

		multiselect({ message, options });

		expect(opt(option, 'active')).toBe(`${color.cyan(S_CHECKBOX_ACTIVE)} ${option.label}`);
		expect(opt(option, 'active-selected')).toBe(
			`${color.green(S_CHECKBOX_SELECTED)} ${option.label}`
		);
		expect(opt(option, 'cancelled')).toBe(`${color.strikethrough(color.dim(option.label))}`);
		expect(opt(option, 'inactive')).toBe(
			`${color.dim(S_CHECKBOX_INACTIVE)} ${color.dim(option.label)}`
		);
		option.label = undefined;
		expect(opt(option, 'inactive')).toBe(
			`${color.dim(S_CHECKBOX_INACTIVE)} ${color.dim(option.value)}`
		);
	});

	it('should render initial state', () => {
		multiselect({ message, options });

		expect(mock.state).toBe('initial');
		expect(mock.frame).toMatchSnapshot();
	});

	it('should render initial state with initialValue', () => {
		multiselect({
			message,
			options,
			initialValues: [options[0].value, options[1].value],
			cursorAt: options[1].value,
		});

		expect(mock.state).toBe('initial');
		expect(mock.frame).toMatchSnapshot();
	});

	it('should render error', () => {
		multiselect({ message, options });
		mock.submit();

		expect(mock.state).toBe('error');
		expect(mock.frame).toMatchSnapshot();
	});

	it('should render cancel', () => {
		const title = `${color.gray(S_BAR)}\n${symbol('cancel')}  ${message}\n`;

		multiselect({ message, options });
		mock.cancel();

		expect(mock.state).toBe('cancel');
		expect(mock.frame).toBe(`${title}${color.gray(S_BAR)}`);
	});

	it('should render cancel with value', () => {
		const title = `${color.gray(S_BAR)}\n${symbol('cancel')}  ${message}\n`;

		multiselect({ message, options });
		mock.cancel([options[0].value, options[1].value]);

		expect(mock.state).toBe('cancel');
		expect(mock.frame).toBe(
			`${title}${color.gray(S_BAR)}  ${[options[0], options[1]]
				.map((option) => opt(option, 'cancelled'))
				.join(color.dim(', '))}\n${color.gray(S_BAR)}`
		);
	});

	it('should submit value', () => {
		const title = `${color.gray(S_BAR)}\n${symbol('submit')}  ${message}\n`;

		multiselect({ message, options });
		mock.submit([options[0].value, options[1].value]);

		expect(mock.state).toBe('submit');
		expect(mock.frame).toBe(
			`${title}${color.gray(S_BAR)}  ${color.dim(options[0].value)}${color.dim(', ')}${color.dim(
				options[1].value
			)}`
		);
	});

	it('should return value on submit', async () => {
		const promise = multiselect({ message, options });

		mock.submit([options[0].value, options[1].value]);
		const result = await promise;

		expect(result).toStrictEqual([options[0].value, options[1].value]);
	});

	it('should render limited options', async () => {
		const extendedOptions = options.concat(options);
		const cursor = 0;

		multiselect({ message, options: extendedOptions, maxItems: 5 });
		mock.setCursor(cursor);
		mock.updateFrame();

		const INACTIVE_CHECKBOX = color.dim(S_CHECKBOX_INACTIVE);
		const ACTIVE_CHECKBOX = color.cyan(S_CHECKBOX_ACTIVE);
		expect(mock.frame).toBe(
			[
				color.gray(S_BAR),
				`${symbol('initial')}  ${message}`,
				`${color.cyan(S_BAR)}  ${ACTIVE_CHECKBOX} ${extendedOptions[cursor].value}`,
				`${color.cyan(S_BAR)}  ${INACTIVE_CHECKBOX} ${color.dim(
					extendedOptions[cursor + 1].value
				)}`,
				`${color.cyan(S_BAR)}  ${INACTIVE_CHECKBOX} ${color.dim(
					extendedOptions[cursor + 2].value
				)}`,
				`${color.cyan(S_BAR)}  ${INACTIVE_CHECKBOX} ${color.dim(
					extendedOptions[cursor + 3].value
				)}`,
				`${color.cyan(S_BAR)}  ${color.dim('...')}`,
				color.cyan(S_BAR_END),
				'',
			].join('\n')
		);
	});
});
