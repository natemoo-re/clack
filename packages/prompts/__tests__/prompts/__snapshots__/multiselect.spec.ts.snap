// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiselect should render error 1`] = `
"[90mâ”‚[39m
[33mâ–²[39m  test message
[33mâ”‚[39m  [36mâ—»[39m foo
[33mâ”‚[39m  [2mâ—»[22m [2mbar[22m
[33mâ”‚[39m  [2mâ—»[22m [2mbaz[22m
[33mâ””[39m  [33mPlease select at least one option.[39m
[8m-[28m  [0m[2mPress [90m[47m[7m space [27m[49m[39m to select, [90m[47m[7m enter [27m[49m[39m to submit[22m[0m
"
`;

exports[`multiselect should render initial state 1`] = `
"[90mâ”‚[39m
[36mâ—†[39m  test message
[36mâ”‚[39m  [36mâ—»[39m foo
[36mâ”‚[39m  [2mâ—»[22m [2mbar[22m
[36mâ”‚[39m  [2mâ—»[22m [2mbaz[22m
[36mâ””[39m
"
`;

exports[`multiselect should render initial state with initialValue 1`] = `
"[90mâ”‚[39m
[36mâ—†[39m  test message
[36mâ”‚[39m  [32mâ—¼[39m [2mfoo[22m
[36mâ”‚[39m  [32mâ—¼[39m bar
[36mâ”‚[39m  [2mâ—»[22m [2mbaz[22m
[36mâ””[39m
"
`;
